dot/parachain/dispute/backend.go

```diff
@@ -0,0 +1,207 @@
+package dispute
+
+import (
+	"sync"
+	"time"
+
+	"github.com/ChainSafe/gossamer/dot/parachain/dispute/types"
+	parachainTypes "github.com/ChainSafe/gossamer/dot/parachain/types"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/dgraph-io/badger/v4"
+	"github.com/tidwall/btree"
+)
+
+// Backend is the backend for the dispute coordinator module.
+type Backend interface {
+	// GetEarliestSession returns the earliest session index, if any.
+	GetEarliestSession() (*parachainTypes.SessionIndex, error)
+	// GetRecentDisputes returns the recent disputes, if any.
+	GetRecentDisputes() (*btree.BTree, error)
+	// GetCandidateVotes returns the votes for the given candidate for the specific session-candidate pair, if any.
+	GetCandidateVotes(session parachainTypes.SessionIndex, candidateHash common.Hash) (*types.CandidateVotes, error)
+
+	// SetEarliestSession sets the earliest session index.
+	SetEarliestSession(session *parachainTypes.SessionIndex) error
+	// SetRecentDisputes sets the recent disputes.
+	SetRecentDisputes(recentDisputes *btree.BTree) error
+	// SetCandidateVotes sets the votes for the given candidate for the specific session-candidate pair.
+	SetCandidateVotes(session parachainTypes.SessionIndex, candidateHash common.Hash, votes *types.CandidateVotes) error
+}
+
+// OverlayBackend is the overlay backend for the dispute coordinator module.
+type OverlayBackend interface {
+	Backend
+
+	// WriteToDB writes the given dispute to the database.
+	WriteToDB() error
+	// GetActiveDisputes returns the active disputes.
+	GetActiveDisputes(now int64) (*btree.BTree, error)
+}
+
+// DBBackend is the backend for the dispute coordinator module that uses a database.
+type DBBackend interface {
+	Backend
+
+	// Write writes the given data to the database.
+	Write(earliestSession *parachainTypes.SessionIndex,
+		recentDisputes *btree.BTree,
+		candidateVotes map[types.Comparator]*types.CandidateVotes) error
+}
+
+type syncedEarliestSession struct {
+	*sync.RWMutex
+	*parachainTypes.SessionIndex
+}
+
+func newSyncedEarliestSession() syncedEarliestSession {
+	return syncedEarliestSession{
+		RWMutex: new(sync.RWMutex),
+	}
+}
+
+type syncedRecentDisputes struct {
+	*sync.RWMutex
+	*btree.BTree
+}
```

Comment:
I would suggest using a property name to hold those types instead of embedding them in the struct

```suggestion
type syncedRecentDisputes struct {
	mtx *sync.RWMutex
	btree *btree.BTree
}
```

Reply:
Resolving this to keep the conversation in one place.